pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,998 @@
+			(pool_data.last_weights.0 .1, pool_data.last_weights.1 .1)
+		} else {
+			(pool_data.last_weights.1 .1, pool_data.last_weights.0 .1)
+		}
+	}
+}
+
+pub trait AssetPairPoolIdFor<AssetId: Sized, PoolId: Sized> {
+	fn from_assets(asset_a: AssetId, asset_b: AssetId) -> PoolId;
+}
+
+pub struct AssetPairPoolId<T: Config>(PhantomData<T>);
+
+impl<T: Config> AssetPairPoolIdFor<AssetId, PoolId<T>> for AssetPairPoolId<T>
+where
+	PoolId<T>: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+{
+	fn from_assets(asset_a: AssetId, asset_b: AssetId) -> PoolId<T> {
+		let mut buf: Vec<u8> = b"lbp".to_vec();
+
+		if asset_a < asset_b {
+			buf.extend_from_slice(&asset_a.to_le_bytes());
+			buf.extend_from_slice(&asset_b.to_le_bytes());
+		} else {
+			buf.extend_from_slice(&asset_b.to_le_bytes());
+			buf.extend_from_slice(&asset_a.to_le_bytes());
+		}
+		PoolId::<T>::unchecked_from(T::Hashing::hash(&buf[..]))
+	}
+}
+
+impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, BalanceOf<T>> for Pallet<T> {
+	fn exists(assets: AssetPair) -> bool {
+		let pair_account = T::AssetPairPoolId::from_assets(assets.asset_in, assets.asset_out);
+		<PoolAssets<T>>::contains_key(&pair_account)
+	}
+
+	fn get_pair_id(assets: AssetPair) -> T::AccountId {
+		T::AssetPairPoolId::from_assets(assets.asset_in, assets.asset_out)
+	}
+
+	fn get_pool_assets(pool_account_id: &T::AccountId) -> Option<Vec<AssetId>> {
+		match <PoolAssets<T>>::contains_key(pool_account_id) {
+			true => {
+				let assets = Self::pool_assets(pool_account_id);
+				Some(vec![assets.0, assets.1])
+			}
+			false => None,
+		}
+	}
+
+	/// Calculate spot price for given assets and amount and update pool weights if necessary
+	///
+	/// Provided assets must exist in the pool. Panic if an asset does not exist in the pool.
+	///
+	/// Return 0 if calculation overflows or weights calculation overflows.
+	fn get_spot_price_unchecked(asset_a: AssetId, asset_b: AssetId, amount: BalanceOf<T>) -> BalanceOf<T> {
+		let pool_id = Self::get_pair_id(AssetPair {
+			asset_in: asset_a,
+			asset_out: asset_b,
+		});
+
+		let asset_a_reserve = T::MultiCurrency::free_balance(asset_a, &pool_id);
+		let asset_b_reserve = T::MultiCurrency::free_balance(asset_b, &pool_id);
+
+		// Note : unchecked function; assumes that existence was checked prior to this call
+		let mut pool_data = Self::pool_data(&pool_id);
+
+		let weight_a: LBPWeight;
+		let weight_b: LBPWeight;
+		// update weight or reuse last if update is not necessary
+		if pool_data.last_weight_update == <frame_system::Pallet<T>>::block_number() {
+			let w = Self::get_weights_in_order(&pool_data, asset_a);
+
+			weight_a = w.0;
+			weight_b = w.1;
+		} else {
+			match Self::update_weights(&mut pool_data) {
+				Ok(weights) => {
+					if weights.0 .0 == asset_a {
+						weight_a = weights.0 .1;
+						weight_b = weights.1 .1;
+					} else {
+						weight_a = weights.1 .1;
+						weight_b = weights.0 .1;
+					}
+
+					<PoolData<T>>::insert(&pool_id, &pool_data);
```

Comment:
I wonder why this function changes the pool? 

it should only calculate spot price.

I dont think that this should make storage changes - if yes, it probably needs some refactoring.

Reply:
Calculated weights are stored so there is no need to calculate them again in the same block. I have moved the _insert_ method to _update_weights_ function.